src/error.rs

```diff
@@ -0,0 +1,107 @@
+use std::error::Error;
+use std::fmt;
+use std::io;
+use color::ColorType;
+use image::ImageFormat;
+
+/// An enumeration of Image errors
+#[derive(Debug)]
+pub enum ImageError {
+    /// The Image is not formatted properly
+    FormatError(String),
+
+    /// The Image's dimensions are either too small or too large
+    DimensionError,
+
+    /// This image format is not supported
+    UnsupportedFormat(String),
+
+    /// This color type is not supported
+    UnsupportedColor(ColorType),
+
+    /// Unsupported feature in a image format
+    UnsupportedFeature(ImageFormat, String),
```

Comment:
I'm not sure if we should include `ImageFormat` here. The problem is that if encoding/decoding crates were to start returning `ImageError`s instead of their own error types, then we'd have to add a variant to ImageFormat before we could start implementing the decoding library (so that it could produce correctly typed UnsupportedFeature errors)

Reply:
Maybe this could be a (struct) variant similar to the following. Preferring struct to give the `Option` a better meaning.
```
UnsupportedFeature {
    format: Option<ImageFormat>,
    details: String,
}
```

Reply:
Hm, I'm imagining `ImaegError` becoming part of image-core with `ImageFormat` remaining part of the main image crate. That would preclude mentioning ImageFormat at all, even inside an option. Also, wouldn't we expect the user to have an idea of which image type their decoding anyway? After all, most of the other variants don't list which image format...

Reply:
True, maybe we should avoid mentioning it altogether.

Is `String` the best representation we have for unsupported features? `Cow<'static, str>` seems like a better representation even if we stringly type this variant.

Reply:
This is probably better. I think we have to stringly type this variant as it can be so many different things from any of the decoders, however I don't know the answer to the `Cow` vs `String` question, I see both appear here and there. I would say that the small amount of allocations that will happen using string is unproblematic and the small complexity of using `Cow` isn't really worth it.